cpp/include/raft/spatial/knn/detail/ball_cover.cuh

```diff
@@ -0,0 +1,342 @@
+  rng.sampleWithoutReplacement(handle, R_indices.data(), R_1nn_cols2.data(),
+                               index.get_R_1nn_cols(), R_1nn_ones.data(),
+                               (value_idx)index.n_landmarks, (value_idx)index.m,
+                               handle.get_stream());
+
+  raft::matrix::copyRows<value_t, value_idx, size_t>(
+    index.get_X(), index.m, index.n, index.get_R(), R_1nn_cols2.data(),
+    index.n_landmarks, handle.get_stream(), true);
+}
+
+/**
+ * Constructs a 1-nn index mapping each landmark to their closest points.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param handle
+ * @param R_knn_inds_ptr
+ * @param R_knn_dists_ptr
+ * @param k
+ * @param index
+ */
+template <typename value_idx, typename value_t>
+void construct_landmark_1nn(const raft::handle_t &handle,
+                            const value_idx *R_knn_inds_ptr,
+                            const value_t *R_knn_dists_ptr, int k,
+                            BallCoverIndex<value_idx, value_t> &index) {
+  rmm::device_uvector<value_idx> R_1nn_inds(index.m, handle.get_stream());
+
+  value_idx *R_1nn_inds_ptr = R_1nn_inds.data();
+  value_t *R_1nn_dists_ptr = index.get_R_1nn_dists();
+
+  auto idxs = thrust::make_counting_iterator<value_idx>(0);
+  thrust::for_each(handle.get_thrust_policy(), idxs, idxs + index.m,
+                   [=] __device__(value_idx i) {
+                     R_1nn_inds_ptr[i] = R_knn_inds_ptr[i * k];
+                     R_1nn_dists_ptr[i] = R_knn_dists_ptr[i * k];
+                   });
+
+  auto keys = thrust::make_zip_iterator(
+    thrust::make_tuple(R_1nn_inds.data(), index.get_R_1nn_dists()));
+
+  // group neighborhoods for each reference landmark and sort each group by distance
+  thrust::sort_by_key(handle.get_thrust_policy(), keys, keys + index.m,
+                      index.get_R_1nn_cols(), NNComp());
+
+  // convert to CSR for fast lookup
+  raft::sparse::convert::sorted_coo_to_csr(
+    R_1nn_inds.data(), index.m, index.get_R_indptr(), index.n_landmarks + 1,
+    handle.get_stream());
+}
+
+/**
+ * Computes the k closest landmarks to a set of query points.
+ * @tparam value_idx
+ * @tparam value_t
+ * @tparam value_int
+ * @param handle
+ * @param index
+ * @param query_pts
+ * @param n_query_pts
+ * @param k
+ * @param R_knn_inds
+ * @param R_knn_dists
+ */
+template <typename value_idx, typename value_t, typename value_int = int>
+void k_closest_landmarks(const raft::handle_t &handle,
+                         BallCoverIndex<value_idx, value_t> &index,
+                         const value_t *query_pts, value_int n_query_pts, int k,
+                         value_idx *R_knn_inds, value_t *R_knn_dists) {
+  std::vector<value_t *> input = {index.get_R()};
+  std::vector<int> sizes = {index.n_landmarks};
+
+  brute_force_knn_impl<int, int64_t>(
+    input, sizes, index.n, const_cast<value_t *>(query_pts), n_query_pts,
+    R_knn_inds, R_knn_dists, k, handle.get_stream(), nullptr, 0, (bool)true,
```

Comment:
IIRC, this is a silly artifact from type templating which is meant to avoid compiler confusion about `nvbool`. It's pretty harmless. I was trying to keep from having to explicitly assign the value to a variable. I can try other ways to solve this if you know any.

Reply:
Oh! Gotcha. In that case, I'd just do a `static_cast` and call it a day.